crates/ra_ide/src/references.rs

```diff
@@ -122,8 +122,33 @@ pub(crate) fn find_all_refs(
     let mut search_kind = ReferenceKind::Other;
 
     if let TokenAtOffset::Between(ref left, ref right) = token {
-        if (right.kind() == SyntaxKind::L_CURLY || right.kind() == SyntaxKind::L_PAREN)
-            && left.kind() != SyntaxKind::IDENT
+        let opt_name = find_node_at_offset::<ast::Name>(&syntax, left.text_range().start());
+        let opt_name_ref = find_node_at_offset::<ast::NameRef>(&syntax, left.text_range().start());
+
+        if (((right.kind() == SyntaxKind::WHITESPACE // StructDef/RecordLit<|> {
+            && right.next_token().map(|t| t.kind() == SyntaxKind::L_CURLY).unwrap_or(false))
+                || right.kind() == SyntaxKind::L_CURLY) // StructDef/RecordLit <|>{
+            && ((opt_name.is_some()
+                && left
+                    .ancestors()
+                    .find_map(ast::StructDef::cast)
+                    .map(|l| l.name() == opt_name)
+                    .unwrap_or(false))
+                || (opt_name_ref.is_some()
+                    && left
+                        .ancestors()
+                        .find_map(ast::RecordLit::cast)
+                        .and_then(|l| l.path())
+                        .and_then(|p| p.segment())
+                        .map(|p| p.name_ref() == opt_name_ref)
+                        .unwrap_or(false))))
+            || (right.kind() == SyntaxKind::L_PAREN // StructDef<|>(
+                && opt_name.is_some()
+                && left
+                    .ancestors()
+                    .find_map(ast::StructDef::cast)
+                    .map(|l| l.name() == opt_name)
+                    .unwrap_or(false))
```

Comment:
Yeah, I guess it's a strong contender for the biggest expression I've ever seen in an `if` condition. Moving it to a helper function with some early returns would help here I think!